File: sol/1298/1298.cpp
#include <queue>

struct Solution {
    int maxCandies(vector<int>& status, vector<int>& candies, vector<vector<int>>& keys, vector<vector<int>>& containedBoxes, vector<int>& initialBoxes) {
        int n = status.size();
        queue<int> q;
        vector<bool> unopenedBox(n, false), unusedKey(n, false), opened(n, false);
        for (int b : initialBoxes){
            if (status[b]){
                q.push(b);
            } else {
                unopenedBox[b] = true;
            }
        }
        int total = 0;
        while (!q.empty()){
            int b = q.front();
            q.pop();
            if (opened[b]){
                continue;
            }
            opened[b] = true;
            unopenedBox[b] = false;
            total += candies[b];
            for (int nb : keys[b]){
                if (unopenedBox[nb]){
                    q.push(nb);
                } else {
                    unusedKey[nb] = true;
                }
            }
            for (int nb : containedBoxes[b]){
                if (status[nb] || unusedKey[nb]){
                    q.push(nb);
                } else {
                    unopenedBox[nb] = true;
                }
            }
        }
        return total;
    }
};

File: sol/1298/1298.py
from collections import deque

class Solution:
    def maxCandies(self, status: List[int], candies: List[int], keys: List[List[int]], containedBoxes: List[List[int]], initialBoxes: List[int]) -> int:
        n = len(status)
        q = deque()
        unopenedBox = [False] * n
        unusedKey = [False] * n
        opened = [False] * n
        for b in initialBoxes:
            if status[b]:
                q.append(b)
            else:
                unopenedBox[b] = True
        total = 0
        while len(q) != 0:
            b = q.pop()
            if opened[b]:
                continue
            opened[b] = True
            unopenedBox[b] = False
            total += candies[b]
            for nb in keys[b]:
                if unopenedBox[nb]:
                    q.append(nb)
                else:
                    unusedKey[nb] = True
            for nb in containedBoxes[b]:
                if status[nb] or unusedKey[nb]:
                    q.append(nb)
                else:
                    unopenedBox[nb] = True
        return total
